### PR TITLE
feat: support keccak family

### LINKED ISSUE #3: Make sha3 an optional dependency
In some environments sha3 won't compile. We should make it optional and fall back to the pure JavaScript version instead of failing in this case.

### PR DESCRIPTION
**Pull request by dignifiedquire**
This also drops the dependency on the native sha3 implementation which turned out to be slower then the currently used JavaScript version

Fixes #3 

Ref https://github.com/multiformats/js-multihash/issues/20

@diasdavid @kumavis please review this, especially the fixtures if they all match the correct hashes as expected.

The codec changes that were made in multihash are breaking, what was before `sha3` is now `keccak-512` and as such has a different hash. 

Depends on https://github.com/multiformats/js-multihash/pull/22

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -73,7 +73,16 @@ Multihashing.functions = {
...
+  0x14: crypto.sha3512,
+  0x15: crypto.sha3384,
+  0x16: crypto.sha3256,
+  0x17: crypto.sha3224,
+  0x18: crypto.shake128,
+  0x19: crypto.shake256,
+  0x1A: crypto.keccak224,
+  0x1B: crypto.keccak256,
+  0x1C: crypto.keccak384,
+  0x1D: crypto.keccak512
```
Comment (diasdavid) [inline src/index.js]: Could you add a comment per line with the name of the hash (i.e: "SHA3-512"), just to make it easier to read through

**Comment by diasdavid** [comment]
Overall LGTM, I can check the validity of the `hashes`, but I would trust the libraries to have tests that verify their own hashing functions (if they don't have tests, then we need to reconsider).

Related with the update with js-multihash, make sure the varint and the total length of the multihash is right (code + varint + digest)

File: src/index.js (hunk 1 above)
Comment (dignifiedquire) [inline src/index.js]: done

**Comment by dignifiedquire** [comment]
@diasdavid don't need varints, I just can't count :) please check if ready now :)

**Comment by diasdavid** [comment]
@dignifiedquire https://github.com/multiformats/js-multihash/pull/22#discussion_r90025382 you need varints

**Comment by diasdavid** [comment]
You don't need varints for sha functions, as the output is fixed size, that isn't true for all of the others. Nevertheless, I believe I understand what you are saying: "You don't need varints, ** for now**"

**Comment by dignifiedquire** [comment]
> You don't need varints, ** for now**"

yes :)

**Pull request merged**